### ISSUE TITLE
`building_env.py` doesn't work on macOS (anymore?)

### ISSUE THREAD

**Issue by markusfisch** [viaduck/openssl-cmake#20]
When trying to use this on macOS, I ran into a couple of problems:

### 1. CMake 3.27 has dropped support for `PythonInterp` (which is deprecated since 3.12) so `cmake/BuildOpenSSL.cmake` needs to use `find_package(Python 3 REQUIRED)` to work with a recent CMake version (3.30.1 is the current version on [homebrew](https://formulae.brew.sh/formula/cmake)).

<details>
<summary>patch</summary>

```
diff --git a/cmake/BuildOpenSSL.cmake b/cmake/BuildOpenSSL.cmake
index 4cb31b2..357869d 100644
--- a/cmake/BuildOpenSSL.cmake
+++ b/cmake/BuildOpenSSL.cmake
@@ -29,7 +29,7 @@ include(ExternalProject)
 
 # find packages
 find_package(Git REQUIRED)
-find_package(PythonInterp 3 REQUIRED)
+find_package(Python 3 REQUIRED)
 
 # # used to apply various patches to OpenSSL
 find_program(PATCH_PROGRAM patch)
diff --git a/scripts/building_env.py b/scripts/building_env.py
old mode 100644
new mode 100755
```

</details>

### 2. Then, building fails with a "Permission denied" error for `scripts/building_env.py`.

<details>
<summary>log</summary>

```
cmake -S . -B build
-- The C compiler identification is AppleClang 15.0.0.15000309
-- The CXX compiler identification is AppleClang 15.0.0.15000309
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Found Git: /usr/bin/git (found version "2.39.3 (Apple Git-146)")
-- Found Python: /opt/homebrew/Frameworks/Python.framework/Versions/3.12/bin/python3.12 (found suitable version "3.12.4", minimum required is "3") found components: Interpreter
-- Configuring done (0.4s)
-- Generating done (0.0s)
-- Build files have been written to: /Users/markus/openssl-cmake-probe/build
cmake --build build
[  7%] Performing setGitConfig step for 'openssl'
[ 15%] Creating directories for 'openssl'
[ 23%] Performing download step (download, verify and extract) for 'openssl'
-- Downloading...
   dst='/Users/markus/openssl-cmake-probe/build/external/openssl-cmake/openssl-prefix/src/openssl-3.1.5.tar.gz'
   timeout='none'
   inactivity timeout='none'
-- Using src='https://mirror.viaduck.org/openssl/openssl-3.1.5.tar.gz'
-- [download 0% complete]
…
-- [download 100% complete]
-- Downloading... done
-- extracting...
     src='/Users/markus/openssl-cmake-probe/build/external/openssl-cmake/openssl-prefix/src/openssl-3.1.5.tar.gz'
     dst='/Users/markus/openssl-cmake-probe/build/external/openssl-cmake/openssl-prefix/src/openssl'
-- extracting... [tar xfz]
-- extracting... [analysis]
-- extracting... [rename]
-- extracting... [clean up]
-- extracting... done
[ 30%] Performing restoreGitConfig step for 'openssl'

[ 38%] No update step for 'openssl'
[ 46%] Performing patch step for 'openssl'

patching file 'test/recipes/80-test_cms.t'
No such line 1143 in input file, ignoring
Ignoring previously applied (or reversed) patch.
1 out of 1 hunks ignored--saving rejects to -
@@ -1052,9 +1052,13 @@
 # Test case for return value mis-check reported in #21986
 with({ exit_checker => sub { return shift == 3; } },
     sub {
-        ok(run(app(['openssl', 'cms', '-sign',
-                    '-in', srctop_file("test", "smcont.txt"),
-                    '-signer', srctop_file("test/smime-certs", "smdsa1.pem"),
-                    '-md', 'SHAKE256'])),
-           "issue#21986");
+        SKIP: {
+          skip "DSA is not supported in this build", 1 if $no_dsa;
+
+          ok(run(app(['openssl', 'cms', '-sign',
+                      '-in', srctop_file("test", "smcont.txt"),
+                      '-signer', srctop_file("test/smime-certs", "smdsa1.pem"),
+                      '-md', 'SHAKE256'])),
+            "issue#21986");
+        }
     });

[ 53%] Performing configure step for 'openssl'
/bin/sh: /Users/markus/openssl-cmake-probe/external/openssl-cmake/scripts/building_env.py: Permission denied
make[3]: *** [external/openssl-cmake/openssl-prefix/src/openssl-stamp/openssl-configure] Error 126
make[2]: *** [external/openssl-cmake/CMakeFiles/openssl.dir/all] Error 2
make[1]: *** [all] Error 2
make: *** [build/foo] Error 2
```

</details>

Fixing this with:
```
$ chmod a+x /Users/markus/openssl-cmake-probe/external/openssl-cmake/scripts/building_env.py
```

### 3. Then, `building_env.py` fails with syntax errors:

<details>
<summary>log</summary>

```
cmake -S . -B build
-- The C compiler identification is AppleClang 15.0.0.15000309
-- The CXX compiler identification is AppleClang 15.0.0.15000309
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Found Git: /usr/bin/git (found version "2.39.3 (Apple Git-146)")
-- Found Python: /opt/homebrew/Frameworks/Python.framework/Versions/3.12/bin/python3.12 (found suitable version "3.12.4", minimum required is "3") found components: Interpreter
-- Configuring done (0.4s)
-- Generating done (0.0s)
-- Build files have been written to: /Users/markus/openssl-cmake-probe/build
cmake --build build
[  7%] Performing setGitConfig step for 'openssl'
[ 15%] Creating directories for 'openssl'
[ 23%] Performing download step (download, verify and extract) for 'openssl'
-- Downloading...
   dst='/Users/markus/openssl-cmake-probe/build/external/openssl-cmake/openssl-prefix/src/openssl-3.1.5.tar.gz'
   timeout='none'
   inactivity timeout='none'
-- Using src='https://mirror.viaduck.org/openssl/openssl-3.1.5.tar.gz'
-- [download 0% complete]
…
-- [download 100% complete]
-- Downloading... done
-- extracting...
     src='/Users/markus/openssl-cmake-probe/build/external/openssl-cmake/openssl-prefix/src/openssl-3.1.5.tar.gz'
     dst='/Users/markus/openssl-cmake-probe/build/external/openssl-cmake/openssl-prefix/src/openssl'
-- extracting... [tar xfz]
-- extracting... [analysis]
-- extracting... [rename]
-- extracting... [clean up]
-- extracting... done
[ 30%] Performing restoreGitConfig step for 'openssl'

[ 38%] No update step for 'openssl'
[ 46%] Performing patch step for 'openssl'

patching file 'test/recipes/80-test_cms.t'
No such line 1143 in input file, ignoring
Ignoring previously applied (or reversed) patch.
1 out of 1 hunks ignored--saving rejects to -
@@ -1052,9 +1052,13 @@
 # Test case for return value mis-check reported in #21986
 with({ exit_checker => sub { return shift == 3; } },
     sub {
-        ok(run(app(['openssl', 'cms', '-sign',
-                    '-in', srctop_file("test", "smcont.txt"),
-                    '-signer', srctop_file("test/smime-certs", "smdsa1.pem"),
-                    '-md', 'SHAKE256'])),
-           "issue#21986");
+        SKIP: {
+          skip "DSA is not supported in this build", 1 if $no_dsa;
+
+          ok(run(app(['openssl', 'cms', '-sign',
+                      '-in', srctop_file("test", "smcont.txt"),
+                      '-signer', srctop_file("test/smime-certs", "smdsa1.pem"),
+                      '-md', 'SHAKE256'])),
+            "issue#21986");
+        }
     });

[ 53%] Performing configure step for 'openssl'
Version: ImageMagick 7.1.1-36 Q16-HDRI aarch64 22352 https://imagemagick.org
Copyright: (C) 1999 ImageMagick Studio LLC
License: https://imagemagick.org/script/license.php
Features: Cipher DPC HDRI Modules OpenMP(5.0)
Delegates (built-in): bzlib fontconfig freetype gslib heic jng jp2 jpeg jxl lcms lqr ltdl lzma openexr png ps raw tiff webp xml zlib zstd
Compiler: gcc (4.2)
Usage: import [options ...] [ file ]

Image Settings:
  -adjoin              join images into a single multi-image file
  -border              include window border in the output image
  -channel type        apply option to select image channels
  -colorspace type     alternate image colorspace
  -comment string      annotate image with comment
  -compress type       type of pixel compression when writing the image
  -define format:option
                       define one or more image format options
  -density geometry    horizontal and vertical density of the image
  -depth value         image depth
  -descend             obtain image by descending window hierarchy
  -display server      X server to contact
  -dispose method      layer disposal method
  -dither method       apply error diffusion to image
  -delay value         display the next image after pausing
  -encipher filename   convert plain pixels to cipher pixels
  -endian type         endianness (MSB or LSB) of the image
  -encoding type       text encoding type
  -filter type         use this filter when resizing an image
  -format "string"     output formatted image characteristics
  -frame               include window manager frame
  -gravity direction   which direction to gravitate towards
  -identify            identify the format and characteristics of the image
  -interlace type      None, Line, Plane, or Partition
  -interpolate method  pixel color interpolation method
  -label string        assign a label to an image
  -limit type value    Area, Disk, Map, or Memory resource limit
  -monitor             monitor progress
  -page geometry       size and location of an image canvas
  -pause seconds       seconds delay between snapshots
  -pointsize value     font point size
  -quality value       JPEG/MIFF/PNG compression level
  -quiet               suppress all warning messages
  -regard-warnings     pay attention to warning messages
  -repage geometry     size and location of an image canvas
  -respect-parentheses settings remain in effect until parenthesis boundary
  -sampling-factor geometry
                       horizontal and vertical sampling factor
  -scene value         image scene number
  -screen              select image from root window
  -seed value          seed a new sequence of pseudo-random numbers
  -set property value  set an image property
  -silent              operate silently, i.e. don't ring any bells
  -snaps value         number of screen snapshots
  -support factor      resize support: > 1.0 is blurry, < 1.0 is sharp
  -synchronize         synchronize image to storage device
  -taint               declare the image as modified
  -transparent-color color
                       transparent color
  -treedepth value     color tree depth
  -verbose             print detailed information about the image
  -virtual-pixel method
                       Constant, Edge, Mirror, or Tile
  -window id           select window with this id or name
                       root selects whole screen

Image Operators:
  -annotate geometry text
                       annotate the image with text
  -colors value        preferred number of colors in the image
  -crop geometry       preferred size and location of the cropped image
  -encipher filename   convert plain pixels to cipher pixels
  -extent geometry     set the image size
  -geometry geometry   preferred size or location of the image
  -help                print program options
  -monochrome          transform image to black and white
  -negate              replace every pixel with its complementary color
  -quantize colorspace reduce colors in this colorspace
  -resize geometry     resize the image
  -rotate degreeimport: delegate library support not built-in '' (X11) @ error/import.c/ImportImageCommand/1302.
s      apply Paeth rotation to the image
  -strip               strip image of all profiles and comments
  -thumbnail geometry  create a thumbnail of the image
  -transparent color   make this color transparent within the image
  -trim                trim image edges
  -type type           image type

Miscellaneous Options:
  -debug events        display copious debugging information
  -help                print program options
  -list type           print a list of supported option arguments
  -log format          format of debugging information
  -version             print version information

By default, 'file' is written in the MIFF image format.  To
specify a particular image format, precede the filename with an image
format name and a colon (i.e. ps:image) or specify the image type as
the filename suffix (i.e. image.ps).  Specify 'file' as '-' for
standard input or output.
Version: ImageMagick 7.1.1-36 Q16-HDRI aarch64 22352 https://imagemagick.org
Copyright: (C) 1999 ImageMagick Studio LLC
License: https://imagemagick.org/script/license.php
Features: Cipher DPC HDRI Modules OpenMP(5.0)
Delegates (built-in): bzlib fontconfig freetype gslib heic jng jp2 jpeg jxl lcms lqr ltdl lzma openexr png ps raw tiff webp xml zlib zstd
Compiler: gcc (4.2)
Usage: import [options ...] [ file ]

Image Settings:
  -adjoin              join images into a single multi-image file
  -border              include window border in the output image
  -channel type        apply option to select image channels
  -colorspace type     alternate image colorspace
  -comment string      annotate image with comment
  -compress type       type of pixel compression when writing the image
  -define format:option
                       define one or more image format options
  -density geometry    horizontal and vertical density of the image
  -depth value         image depth
  -descend             obtain image by descending window hierarchy
  -display server      X server to contact
  -dispose method      layer disposal method
  -dither method       apply error diffusion to image
  -delay value         display the next image after pausing
  -encipher filename   convert plain pixels to cipher pixels
  -endian type         endianness (MSB or LSB) of the image
  -encoding type       text encoding type
  -filter type         use this filter when resizing an image
  -format "string"     output formatted image characteristics
  -frame               include window manager frame
  -gravity direction   which direction to gravitate towards
  -identify            identify the format and characteristics of the image
  -interlace type      None, Line, Plane, or Partition
  -interpolate method  pixel color interpolation method
  -label string        assign a label to an image
  -limit type value    Area, Disk, Map, or Memory resource limit
  -monitor             monitor progress
  -page geometry       size and location of an image canvas
  -pause seconds       seconds delay between snapshots
  -pointsize value     font point size
  -quality value       JPEG/MIFF/PNG compression level
  -quiet               suppress all warning messages
  -regard-warnings     pay attention to warning messages
  -repage geometry     size and location of an image canvas
  -respect-parentheses settings remain in effect until parenthesis boundary
  -sampling-factor geometry
                       horizontal and vertical sampling factor
  -scene value         image scene number
  -screen              select image from root window
  -seed value          seed a new sequence of pseudo-random numbers
  -set property value  set an image property
  -silent              operate silently, i.e. don't ring any bells
  -snaps value         number of screen snapshots
  -support factor      resize support: > 1.0 is blurry, < 1.0 is sharp
  -synchronize         synchronize image to storage device
  -taint               declare the image as modified
  -transparent-color color
                       transparent color
  -treedepth value     color tree depth
  -verbose             print detailed information about the image
  -virtual-pixel method
                       Constant, Edge, Mirror, or Tile
  -window id           select window with this id or name
                       root selects whole screen

Image Operators:
  -annotate geometry text
                       annotate the image with text
  -colors value        preferred number of colors in the image
  -crop geometry       preferred size and location of the cropped image
  -encipher filename   convert plain pixels to cipher pixels
  -extent geometry     set the image size
  -geometry geometry   preferred size or location of the image
  -help                print program options
  -monochrome          transform image to black and white
  -negate              replace every pixel with its complementary color
  -quantize colorspace reduce colors in this colorspace
  -resize geometry     resize the image
  -rotate degreeimport: delegate library support not built-in '' (X11) @ error/import.c/ImportImageCommand/1302.
s      apply Paeth rotation to the image
  -strip               strip image of all profiles and comments
  -thumbnail geometry  create a thumbnail of the image
  -transparent color   make this color transparent within the image
  -trim                trim image edges
  -type type           image type

Miscellaneous Options:
  -debug events        display copious debugging information
  -help                print program options
  -list type           print a list of supported option arguments
  -log format          format of debugging information
  -version             print version information

By default, 'file' is written in the MIFF image format.  To
specify a particular image format, precede the filename with an image
format name and a colon (i.e. ps:image) or specify the image type as
the filename suffix (i.e. image.ps).  Specify 'file' as '-' for
standard input or output.
/Users/markus/openssl-cmake-probe/external/openssl-cmake/scripts/building_env.py: line 30: from: command not found
/Users/markus/openssl-cmake-probe/external/openssl-cmake/scripts/building_env.py: line 31: from: command not found
/Users/markus/openssl-cmake-probe/external/openssl-cmake/scripts/building_env.py: line 33: syntax error near unexpected token `('
/Users/markus/openssl-cmake-probe/external/openssl-cmake/scripts/building_env.py: line 33: `parser = argparse.ArgumentParser()'
make[3]: *** [external/openssl-cmake/openssl-prefix/src/openssl-stamp/openssl-configure] Error 2
make[2]: *** [external/openssl-cmake/CMakeFiles/openssl.dir/all] Error 2
make[1]: *** [all] Error 2
make: *** [build/foo] Error 2
```

</details>

I am running Python 3.12.4 (from homebrew) and also tried version 3.9.6 that comes with macOS.

Here is the `CMakeLists.txt` and `main.cpp` I have used to make this easily reproducible:

<details>
<summary>CMakeLists.txt</summary>

```
cmake_minimum_required(VERSION 3.19)

project(Foo C CXX)

set(CMAKE_CXX_STANDARD 20)

set(BUILD_OPENSSL ON)
set(OPENSSL_BUILD_VERSION 3.1.5)
add_subdirectory(external/openssl-cmake)

add_executable(foo main.cpp)
target_include_directories(foo PUBLIC ${OPENSSL_INCLUDE_DIR})
target_link_libraries(foo PRIVATE ssl crypto)
```

</details>
<details>
<summary>main.cpp</summary>

```
#include <openssl/bio.h>

int main(int argc, char **argv) {
	BIO_free(BIO_new(BIO_s_mem()));
	return 0;
}
```

</details>

**Comment by kleest** [comment]
Thanks for the bug report, working on it.

**Comment by kleest** [comment]
Should be fixed by 0e04e101507df752500093ff4b04cce5a898abaa. Please test.

**Comment by markusfisch** [comment]
Yes, it's working now! Thanks a lot for the quick fix! 👍
